sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java

```diff
@@ -0,0 +1,977 @@
+      PCollectionTuple writeResults =
+          input.apply(
+              "Write nd json to GCS",
+              ParDo.of(new WriteBundlesToFilesFn(fhirStore, tempGcsPath, deadLetterGcsPath))
+                  .withOutputTags(Write.TEMP_FILES, TupleTagList.of(Write.FAILED_BODY)));
+
+      int numShards = 100;
+      PCollection<HealthcareIOError<String>> importResults =
+          writeResults
+              .get(Write.TEMP_FILES)
+              .apply(
+                  "Shard files", // to paralelize group into batches
+                  WithKeys.of(ThreadLocalRandom.current().nextInt(0, numShards)))
+              .apply("File Batches", GroupIntoBatches.ofSize(batchSize))
+              .apply(
+                  ParDo.of(
+                      new ImportFn(fhirStore, tempGcsPath, deadLetterGcsPath, contentStructure)))
+              .setCoder(new HealthcareIOErrorCoder<>(StringUtf8Coder.of()));
+
+      return writeResults
+          .get(Write.FAILED_BODY)
+          .setCoder(new HealthcareIOErrorCoder<>(StringUtf8Coder.of()));
+    }
+
+    static class ImportFn
+        extends DoFn<KV<Integer, Iterable<ResourceId>>, HealthcareIOError<String>> {
+
+      private static final Logger LOG = LoggerFactory.getLogger(ImportFn.class);
+      private final String tempGcsPath;
+      private final String deadLetterGcsPath;
+      private ResourceId tempDir;
+      private final ContentStructure contentStructure;
+      private HealthcareApiClient client;
+      private final String fhirStore;
+
+      ImportFn(
+          String fhirStore,
+          String tempGcsPath,
+          String deadLetterGcsPath,
+          @Nullable ContentStructure contentStructure) {
+        this.fhirStore = fhirStore;
+        this.tempGcsPath = tempGcsPath;
+        this.deadLetterGcsPath = deadLetterGcsPath;
+        if (contentStructure == null) {
+          this.contentStructure = ContentStructure.CONTENT_STRUCTURE_UNSPECIFIED;
+        } else {
+          this.contentStructure = contentStructure;
+        }
+      }
+
+      @Setup
+      public void init() throws IOException {
+        tempDir =
+            FileSystems.matchNewResource(tempGcsPath, true)
+                .resolve(
+                    String.format("tmp-%s", UUID.randomUUID().toString()),
+                    StandardResolveOptions.RESOLVE_DIRECTORY);
+        client = new HttpHealthcareApiClient();
+      }
+
+      /**
+       * Move files to a temporary subdir (to provide common prefix) to execute import with single
+       * GCS URI.
+       */
+      @ProcessElement
+      public void importBatch(ProcessContext context) throws IOException {
+        Iterable<ResourceId> batch = context.element().getValue();
+        List<ResourceId> tempDestinations = new ArrayList<>();
+        List<ResourceId> deadLetterDestinations = new ArrayList<>();
+        assert batch != null;
+        for (ResourceId file : batch) {
+          tempDestinations.add(
```

Comment:
what happens if this gets rescheduled.

Reply:
The steps of ImportFn are:

1. [move files](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR765) for this batch import call to a temp prefix (in order to capture with a single wildcard for the import API call) with a [UUID](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR740)
2. [Make and FHIR Import call](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR770)
3. [Poll/block until this operation completes](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR772).
4. [Clean up temp files](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR774)

The FHIR import operation should be idempotent (see comment below).

Reply:
Let's consider what happens if the process get killed / rescheduled between each of these setps
| when?                                         | GCS                                                                                                               | FHIR Store                                                                                                                                | Notes                                                                                                                                        
|-----------------------------------------------|-------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|
| (1) -> (1) between non-atomic object moves on GCS | some objects moved to a tmp-UUID prefix. | No FHIR Import Job Attempted | when rescheduled, will try to import the subset of objects that were NOT moved.
| (1) -> (2) | objects moved to a tmp-UUID prefix. | No FHIR Import Job Attempted | when rescheduled we'll fail because  the original URIs don't exist.                                                                                             
| (2) -> (3)                                    | objects moved to a tmp-UUID prefix                                                                                | an in-flight FHIR Import job that beam has lost track of. We could safely re-do this import job without consequence as it is idempotent.  | However, a rescheduled attempt will fail as the original object URIs now will not exist because they've been "claimed" by being moved to the temp dir. | 
| (3) -> (3) aka killed between polling calls:  | In some cases where InterruptedException is thrown we will move the objects in questions to the dead letter path. | in-flight FHIR Import job may or may not succeed. |                                                                                                                                               If user re-runs import from DLQ should be fine due to idempontency. But this could be wasteful if the original job in-fact succeeds |
| (3) -> (4):                                   | tmp-UUID dir will not be cleaned up properly                                                                      | FHIR Import Job succeeded  |  This is not a critical failure but inconvenient and will incur storage costs. |

Reply:
We can refactor to split this up into:
1. ImportFn
    - Responsible for moving GCS files to temp dir and starting the FHIR Import
1. BlockFn
    - Responsible for blocking on each FHIR Import Operation and cleaning up GCS files
This will help our (2)-> (3), (3) -> (3) and (3) -> (4) scenarios because the OperationID and temp resource IDs will be passed between stages via `PCollection<KV<String, List<ReadableFile>>>`.

This leaves use with the (1) -> (1) and (1) -> (2) scenarios. These could be handled if we "look for" a temp dir similar to the one we'd plan on creating. Aka we look at the existing temp dirs and see if one has matching file names.

Reply:
In 5540277 Refactored FhirIO.Import to:
- copy (instead of rename) files to temp dirs. 
- delete everything from gcsTempDir once window has closed.

This addresses all issues in above table except being killed between polling calls.
To mitigate this we could split into a `ImportFn` that is responsible only for starting Import jobs and  outputs KV of Operation IDs and temp dirs and `BlockTilDoneFn` which handles blocking and clean up.
@pabloem thoughts on a cleaner way to do this?